contracts/src/v0.8/automation/dev/v2_3/AutomationRegistryLogicA2_3.sol

```diff
@@ -302,13 +305,25 @@ contract AutomationRegistryLogicA2_3 is AutomationRegistryBase2_3, Chainable {
    * @param id the upkeepID
    * @param amount the amount of funds to add, in the upkeep's billing token
    */
-  function addFunds(uint256 id, uint96 amount) external {
+  function addFunds(uint256 id, uint96 amount) external payable {
     Upkeep memory upkeep = s_upkeep[id];
     if (upkeep.maxValidBlocknumber != UINT32_MAX) revert UpkeepCancelled();
-    s_upkeep[id].balance = upkeep.balance + amount;
-    s_reserveAmounts[address(upkeep.billingToken)] = s_reserveAmounts[address(upkeep.billingToken)] + amount;
-    bool success = upkeep.billingToken.transferFrom(msg.sender, address(this), amount);
-    if (!success) revert TransferFailed();
+
+    if (msg.value != 0) {
+      // native payment
+      if (upkeep.billingToken != IERC20(i_wrappedNativeTokenAddress)) {
+        revert InvalidBillingToken();
+      }
+      s_upkeep[id].balance = upkeep.balance + uint96(msg.value);
+      IWrappedNative(i_wrappedNativeTokenAddress).deposit{value: msg.value}();
+    } else {
+      // non-native payment
+      s_upkeep[id].balance = upkeep.balance + amount;
+      s_reserveAmounts[address(upkeep.billingToken)] = s_reserveAmounts[address(upkeep.billingToken)] + amount;
```

Comment:
I think we need to pull this out of the `if .. else` so that it happens for both cases :)

Reply:
I think we could pull out the `s_upkeep[id].balance = upkeep.balance + amount;` also for slightly drier code

Reply:
Good catch! for the dry-ness, we want to do the state updates before the actual transfer. that means if we want to reuse the s_reserveAmounts and s_upkeep lines, we have to use 2 if statements, 1st to get the actual amount, 2nd is to make the transfer. I dont think it gets dryer. So I would prefer to keep it as is for fewer branches / better readability.